src/hb-ot-metrics.cc

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright © 2018  Ebrahim Byagowi
+ *
+ *  This is part of HarfBuzz, a text shaping library.
+ *
+ * Permission is hereby granted, without written agreement and without
+ * license or royalty fees, to use, copy, modify, and distribute this
+ * software and its documentation for any purpose, provided that the
+ * above copyright notice and the following two paragraphs appear in
+ * all copies of this software.
+ *
+ * IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE TO ANY PARTY FOR
+ * DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
+ * ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN
+ * IF THE COPYRIGHT HOLDER HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ *
+ * THE COPYRIGHT HOLDER SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING,
+ * BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ * FITNESS FOR A PARTICULAR PURPOSE.  THE SOFTWARE PROVIDED HEREUNDER IS
+ * ON AN "AS IS" BASIS, AND THE COPYRIGHT HOLDER HAS NO OBLIGATION TO
+ * PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
+ */
+
+#include "hb.hh"
+
+#include "hb-ot-gasp-table.hh"
+#include "hb-ot-os2-table.hh"
+#include "hb-ot-post-table.hh"
+#include "hb-ot-hmtx-table.hh"
+#include "hb-ot-var-mvar-table.hh"
+
+#include "hb-ot-face.hh"
+
+#if 0
+static bool
+_get_gasp (hb_font_t *font, unsigned int i, hb_position_t *result,
+	   hb_ot_metrics_t metrics_tag)
+{
+  const OT::GaspRange& range = font->face->table.gasp->get_gasp_range (i);
+  if (&range == &Null (OT::GaspRange)) return false;
+  if (result) *result = font->em_scale_y (range.rangeMaxPPEM +
+					  font->face->table.MVAR->get_var (metrics_tag, nullptr, 0));
+  return true;
+}
+#endif
+
+/**
+ * hb_ot_metrics_get:
+ * @font:
+ * @metrics_tag:
+ * @position:    (out) (optional):
+ *
+ * Returns: Whether found the requested metrics
+ *
+ * Since: REPLACEME
+ **/
+hb_bool_t
+hb_ot_metrics_get (hb_font_t       *font,
+		   hb_ot_metrics_t  metrics_tag,
+		   hb_position_t   *position     /* OUT.  May be NULL. */)
```

Comment:
Ok.  Thought about more of the API.  I like to have this, in this file:

- `hb_ot_metrics_get_position(font, metrics_tag, *position)` for current API,
- `hb_position_t hb_ot_metrics_get_x_variation(font, metrics_tag)`
- `hb_position_t hb_ot_metrics_get_y_variation(font, metrics_tag)`

The current API (get_position) needs to fill in position with sane fallback, such that clients can continue even if it returned false.  We need to pick fallback values one by one.

Re hmtx/vmtx, there's a chicken-and-egg issue.  I'll comment on that separately.

Reply:
The API part is done. Don't know about the fallback part yet...